kmapper/kmapper.py

```diff
@@ -522,6 +530,36 @@ def map(
             self._summary(graph, str(datetime.now() - start))
 
         return graph
+    
+    def _remove_duplicate_nodes(self, nodes):
+        nodes_to_remove = []
+            
+        node_items = list(nodes.items())
+        num_nodes = len(node_items)
+
+        # nested for-loops over ranges for upper-left triangle of pairwise matrix of node_ids       
+        for i in range(num_nodes):
+            left_node_key, left_node_items = node_items[i]
+            left_node_itemset = set(left_node_items)
+            
+            for j in range(i + 1, num_nodes):
+                right_node_key, right_node_items = node_items[j]
+                right_node_itemset = set(right_node_items)
+                
+                if left_node_itemset == right_node_itemset:
+                    nodes_to_remove.append(right_node_key)
+                    
+        nodes_removed = [node_id for node_id in nodes_to_remove if nodes.pop(node_id, None) is not None]
```

Comment:
When would `nodes` not have `node_id` in it?  Is the `if ... is not None` necessary?

Reply:
With this pairwise upper left triangle approach, each node is compared against all other nodes (but only once) to check for a match. It is possible that there is a set of, say, three duplicates. In that case, when looping through, the `nodes.pop(node_id, None)` could have already popped that node_id, meaning that the pop would return `None` because it would not be found. If that happens, I didn't want to append `None` to the `nodes_removed` list, therefore the if is not None.

Popping is safer than `del` because `del` will fail if the key does not exist

Reply:
Although for succinctness the `is not None` could be dropped, leaving just the `if .pop(..., None)`...